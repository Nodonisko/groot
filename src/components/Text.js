// @flow
import * as React from 'react'
import { Text as RNText } from 'react-native'
import renderRule from '../theme/renderRule'
import type {
  Style,
  Color,
  TextAlign,
  FontSize,
  FontWeight,
  LineHeight,
} from '../types'
import R from 'ramda'

type TextPropsType = {|
  children: string | React.Node,
  style?: Style,
  color?: Color,
  size?: FontSize,
  weight?: FontWeight,
  lineHeight?: LineHeight,
  align?: TextAlign,
|}

const defaultProps = {
  color: 'black',
  size: 'body',
  weight: 'regular',
  align: 'auto',
}

const styles = {
  text: ({ color, size, weight, lineHeight, align, style }: TextPropsType) => ({
    color,
    fontSize: size,
    fontWeight: weight,
    lineHeight: lineHeight || size,
    textAlign: align,
    extend: {
      condition: !R.isNil(style),
      style,
    },
  }),
}

const Text = (props: TextPropsType) => (
  <RNText style={renderRule(styles.text, props)}>{props.children}</RNText>
)

Text.defaultProps = defaultProps

export default Text
