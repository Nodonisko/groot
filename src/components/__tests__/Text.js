// @flow
import 'react-native'
import React from 'react'
import Text from '../Text'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

it('renders correctly with only required props', () => {
  const component = <Text>Hello world</Text>
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with color prop', () => {
  const component = <Text color="red">Hello world</Text>
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with font size prop', () => {
  const component = <Text size="title1">Hello world</Text>
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with font size prop and line height prop', () => {
  const component = (
    <Text size="title1" lineHeight="largeTitle">
      Hello world
    </Text>
  )
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with font weight props', () => {
  const component = <Text weight="bold">Hello world</Text>
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with font align props', () => {
  const component = <Text align="center">Hello world</Text>
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with custom style props', () => {
  const style = { fontSize: 88, paddingLeft: 20, color: '#ddd' }
  const component = <Text style={style}>Hello world</Text>
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders correctly with all possible props', () => {
  const style = { fontSize: 88, paddingLeft: 20 }
  const component = (
    <Text
      color="red"
      size="title1"
      lineHeight="largeTitle"
      weight="bold"
      align="center"
      style={style}
    >
      Hello world
    </Text>
  )
  const tree = renderer.create(component).toJSON()
  expect(tree).toMatchSnapshot()
})
