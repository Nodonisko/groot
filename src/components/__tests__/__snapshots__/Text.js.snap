// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all possible props 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "rgb(255, 0, 0)",
      "fontSize": 88,
      "fontWeight": "600",
      "lineHeight": 49,
      "paddingLeft": 20,
      "textAlign": "center",
    }
  }
>
  Hello world
</Text>
`;

exports[`renders correctly with color prop 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "rgb(255, 0, 0)",
      "fontSize": 17,
      "fontWeight": "400",
      "lineHeight": 25,
      "textAlign": "auto",
    }
  }
>
  Hello world
</Text>
`;

exports[`renders correctly with custom style props 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "#ddd",
      "fontSize": 88,
      "fontWeight": "400",
      "lineHeight": 25,
      "paddingLeft": 20,
      "textAlign": "auto",
    }
  }
>
  Hello world
</Text>
`;

exports[`renders correctly with font align props 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "rgb(0, 0, 0)",
      "fontSize": 17,
      "fontWeight": "400",
      "lineHeight": 25,
      "textAlign": "center",
    }
  }
>
  Hello world
</Text>
`;

exports[`renders correctly with font size prop 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "rgb(0, 0, 0)",
      "fontSize": 28,
      "fontWeight": "400",
      "lineHeight": 40,
      "textAlign": "auto",
    }
  }
>
  Hello world
</Text>
`;

exports[`renders correctly with font size prop and line height prop 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "rgb(0, 0, 0)",
      "fontSize": 28,
      "fontWeight": "400",
      "lineHeight": 49,
      "textAlign": "auto",
    }
  }
>
  Hello world
</Text>
`;

exports[`renders correctly with font weight props 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "rgb(0, 0, 0)",
      "fontSize": 17,
      "fontWeight": "600",
      "lineHeight": 25,
      "textAlign": "auto",
    }
  }
>
  Hello world
</Text>
`;

exports[`renders correctly with only required props 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "rgb(0, 0, 0)",
      "fontSize": 17,
      "fontWeight": "400",
      "lineHeight": 25,
      "textAlign": "auto",
    }
  }
>
  Hello world
</Text>
`;
